### PR TITLE
CRD for subnet binding

### PR DESCRIPTION
**Pull request by wenyingd**

### CODE REVIEW

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +kubebuilder:validation:XValidation:rule="!(has(self.targetSubnetSet) && has(self.targetSubnet))",message="Only one of targetSubnetSet or targetSubnet can be specified"
+type SubnetConnectionBindingMapSpec struct {
+	// SubnetName is Subnet name which this binding is attached on.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="subnetName is immutable"
+	SubnetName string `json:"subnetName"`
+	// TargetSubnetSet specifies the target subnetSet which a subnet is bound to.
+	// +kubebuilder:validation:Optional
+	TargetSubnetSet string `json:"targetSubnetSet,omitempty"`
```
Comment (jianjuns) [inline pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:12]: TargetSubnetSetName

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	// TargetSubnet specifies the target subnet which a subnet is bound to.
+	// +kubebuilder:validation:Optional
+	TargetSubnet string `json:"targetSubnet,omitempty"`
```
Comment (jianjuns) [inline pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:15]: Would you support Subnet as the PARENT in 9.0 too?

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +kubebuilder:validation:XValidation:rule="!(has(self.targetSubnetSet) && has(self.targetSubnet))",message="Only one of targetSubnetSet or targetSubnet can be specified"
+type SubnetConnectionBindingMapSpec struct {
+	// SubnetName is Subnet name which this binding is attached on.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="subnetName is immutable"
+	SubnetName string `json:"subnetName"`
+	// TargetSubnetSet specifies the target subnetSet which a subnet is bound to.
```
Comment (jianjuns) [inline pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:10]: Please use SubnetSet, Subnet, etc.

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	// +kubebuilder:validation:Optional
+	TargetSubnetSet string `json:"targetSubnetSet,omitempty"`
+	// TargetSubnet specifies the target subnet which a subnet is bound to.
+	// +kubebuilder:validation:Optional
+	TargetSubnet string `json:"targetSubnet,omitempty"`
+	// VLANTrafficTag is the vlan tag configured in the binding. It can be empty, then the handler will choose a valid
+	// value based on the existing subnet connection binding maps bound on the target subnet or subnetSet.
```
Comment (jianjuns) [inline pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:17]: You meant NSX OP will auto allocate a VLAN ID? I suggest not doing that, at least not for this release.

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+// SubnetConnectionBindingMapStatus defines the observed state of SubnetConnectionBindingMap.
+type SubnetConnectionBindingMapStatus struct {
+	Subnet      string                  `json:"subnet,omitempty"`
+	BindingMaps []ConnectionBindingMaps `json:"bindingMaps,omitempty"`
+	Conditions  []Condition             `json:"conditions,omitempty"`
+}
+
+type ConnectionBindingMaps struct {
```
Comment (jianjuns) [inline pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:30]: If we do not need this if we do not auto-allocate VLAN?

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	TargetSubnetSetName string `json:"targetSubnetSetName,omitempty"`
+	// TargetSubnetName specifies the target Subnet which a Subnet is bound to.
+	// +kubebuilder:validation:Optional
+	TargetSubnetName string `json:"targetSubnetName,omitempty"`
+	// VLANTrafficTag is the vlan tag configured in the binding. It can be empty, then the handler will choose a valid
+	// value based on the existing SubnetConnectionBindingMaps bound on the target Subnet or SubnetSet.
+	// +kubebuilder:validation:Maximum:=4095
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Required
+	VLANTrafficTag int64 `json:"vlanTrafficTag"`
```
Comment (jianjuns) [inline pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:21]: Any reason to use int64, not int32?

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go (hunk 5 above)
Comment (jianjuns) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:35]: I feel it is not necessary.

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go (hunk 6 above)
Comment (wenyingd) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:21]: `int64` is used for VlanTrafficTag in the NSX model `SubnetConnectionBindingMap` in SDK. It is easier to keep consistent.

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go (hunk 5 above)
Comment (wenyingd) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:35]: Sure, I can remove it from the status.

Comment (wenyingd) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:35]: BTW, is it OK for you to show the CHILD VpcSubnet path in the status (https://github.com/vmware-tanzu/nsx-operator/pull/845/files#diff-bb38a914041e9190d7a2ba1e76acbecacac053613997e3957f682cf1c0c1d9b3R27) ? Or we shall show the CHILD Subnet CR name in the status?

Comment (wenyingd) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go]: I see, removed.

**Comment by wenyingd** [comment]
@jianjuns I have addressed your comments, could you help review again?

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+type SubnetConnectionBindingMapSpec struct {
+	// SubnetName is the Subnet name which this SubnetConnectionBindingMap is associated.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="subnetName is immutable"
+	SubnetName string `json:"subnetName"`
+	// TargetSubnetSetName specifies the target SubnetSet which a Subnet is bound to.
+	// +kubebuilder:validation:Optional
+	TargetSubnetSetName string `json:"targetSubnetSetName,omitempty"`
+	// TargetSubnetName specifies the target Subnet which a Subnet is bound to.
+	// +kubebuilder:validation:Optional
+	TargetSubnetName string `json:"targetSubnetName,omitempty"`
```
Comment (lxiaopei) [inline pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:15]: do we support update targetSubnetName or targetSubnetSetName or from targetSubnetSetName to targetSubnetName in spec?

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +kubebuilder:validation:XValidation:rule="!(has(self.targetSubnetSetName) && has(self.targetSubnetName))",message="Only one of targetSubnetSetName or targetSubnetName can be specified"
```
Comment (lxiaopei) [inline pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:5]: does this expression also true if there is no targetSubnetSetName and targetSubnetName?

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go (hunk 7 above)
Comment (wenyingd) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:15]: If following NSX API design, yes, we shall support this update. @jianjuns do you feel we shall reject such modification in CR ?

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go (hunk 8 above)
Comment (wenyingd) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:5]: Good catch, updated.

**Review by lxiaopei — DISMISSED** [review]

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go (hunk 7 above)
Comment (jianjuns) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:15]: I am fine either way. If disallow update is easier to implement we can do that too.

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +kubebuilder:validation:XValidation:rule="has(self.targetSubnetSetName) && !has(self.targetSubnetName) || !has(self.targetSubnetSetName) && has(self.targetSubnetName)",message="Only one of targetSubnetSetName or targetSubnetName can be specified"
+type SubnetConnectionBindingMapSpec struct {
+	// SubnetName is the Subnet name which this SubnetConnectionBindingMap is associated.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="subnetName is immutable"
+	SubnetName string `json:"subnetName"`
+	// TargetSubnetSetName specifies the target SubnetSet which a Subnet is bound to.
```
Comment (jianjuns) [inline pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:10]: Should be "bound to" or "connected too"?

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go (hunk 7 above)
Comment (wenyingd) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:15]: I don't feel it saves much code. I would keep it as it was (support user updating the parent subnet/subnetset) in the CRD and implementation.

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go (hunk 9 above)
Comment (wenyingd) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:10]: Updated with "connected to"

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	// +kubebuilder:validation:Optional
+	TargetSubnetSetName string `json:"targetSubnetSetName,omitempty"`
+	// TargetSubnetName specifies the target Subnet which a Subnet is bound to.
+	// +kubebuilder:validation:Optional
+	TargetSubnetName string `json:"targetSubnetName,omitempty"`
+	// VLANTrafficTag is the vlan tag configured in the binding. Note, the value of VLANTrafficTag should be
```
[the remark this answers is not in the document]
Comment (wenyingd) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:16]: updated.

**Review by jianjuns — COMMENTED** [review]
LGTM

**Comment by wenyingd** [comment]
Rebased on ToT code.

File: pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go (hunk 7 above)
Comment (lxiaopei) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:16]: if targetSubnetSetName is set, the specified SubnetSet CR might do not have any NSX subnet yet. Will it report error in this case?

Comment (wenyingd) [reply pkg/apis/vpc/v1alpha1/subnetconnectionbindingmap_types.go:16]: In the reconciler implementation, I will add check about it, if the NSX VpcSubnet doesn't exist (mapping to the SubnetSet CR), I will add condition to describe the unready message.

**Review by lxiaopei — APPROVED** [review]

**Review by dantingl — APPROVED** [review]

**Pull request merged**